### PR TITLE
Add e2e test for the ReviseReplica hook

### PR DESCRIPTION
**Pull request by iawia002**
Signed-off-by: Xinzhao Xu <z2d@jifangcheng.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Add e2e test case for #1149

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/cc @XiShanYongYe-Chang @RainbowMango

### CODE REVIEW

**Comment by RainbowMango** [comment]
Do you think the `karmada-interpreter-webhook-example` should be deployed separately as described at #1136 ? @lonelyCZ is working on it.

**Comment by lonelyCZ** [comment]
Thanks @iawia002 ,
The general idea is this, but it isn't suggested to create directly resource about `workload example`  in `local-up-karmada.sh`, `workload example` is only belong to e2e. I'm working on it that is expected to be finished today.

**Comment by iawia002** [comment]
OK, I can wait for you and update the PR (just keep my e2e case part).

File: test/e2e/resourceinterpreter_test.go [hunk 1]
```diff
@@ -121,4 +123,74 @@ var _ = ginkgo.Describe("Resource interpreter webhook testing", func() {
...
+								ClusterNames: []string{"member2"},
+							},
+							Weight: 1,
+						},
+						{
+							TargetCluster: policyv1alpha1.ClusterAffinity{
+								ClusterNames: []string{"member3"},
+							},
+							Weight: 1,
+						},
```
Comment (XiShanYongYe-Chang) [inline test/e2e/resourceinterpreter_test.go:167]: How about calculating the replicas like this:

https://github.com/karmada-io/karmada/blob/14487548dd9c4970dad003da7d253c25811f2515/test/e2e/scheduling_test.go#L535-L546

In this way, we can make the logic is independent of the specific cluster name.

I also plan to gradually change other E2E parts that involve specific cluster names to those that are not related to cluster names.

File: test/e2e/resourceinterpreter_test.go [hunk 2]
```diff
@@ -121,4 +123,74 @@ var _ = ginkgo.Describe("Resource interpreter webhook testing", func() {
 			framework.RemovePropagationPolicy(karmadaClient, policy.Namespace, policy.Name)
 		})
 	})
+
+	ginkgo.Context("Workload resource ReviseReplica hook testing", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/resourceinterpreter_test.go:127]: How about rename this context to `InterpreterOperation ReviseReplica testing`. That way, it's consistent with the above.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks!

File: test/e2e/resourceinterpreter_test.go (hunk 1 above)
Comment (iawia002) [reply test/e2e/resourceinterpreter_test.go]: Great advice, I was wondering how to avoid using specific cluster names myself, updated, PTAL

File: test/e2e/resourceinterpreter_test.go (hunk 2 above)
Comment (iawia002) [reply test/e2e/resourceinterpreter_test.go]: Updated.

File: test/e2e/resourceinterpreter_test.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply test/e2e/resourceinterpreter_test.go]: I don't have a specific idea yet. It may be determined based on the cluster type or traverse the existing cluster.
If you have a good idea, we can discuss it :)

Comment (iawia002) [reply test/e2e/resourceinterpreter_test.go]: For my case now, I think the way in the scheduling test is just fine.

**Comment by XiShanYongYe-Chang** [comment]
Uhum... Something wrong.

**Comment by iawia002** [comment]
> Uhum... Something wrong.

Yeah, I'm testing it locally.

**Comment by XiShanYongYe-Chang** [comment]
The `framework.ClusterNames()` will get an empty list outside the `ginkgo.It` block, so the sum weight will be 0.

We need to move the calculate of sum weight into  the `ginkgo.It` block.

**Comment by XiShanYongYe-Chang** [comment]
The reason why the `framework.ClusterNames(https://github.com/karmada-io/karmada/blob/14487548dd9c4970dad003da7d253c25811f2515/test/e2e/framework/cluster.go#L32)` will get an empty, I'm guessing the value of [clusterNames]() is determined at Context init-time, and it hasn't been processed yet, so it's going to be nil.

When running `ginkgo.It`, the `clusterNames` will be calculated, so it will be ok.

**Comment by iawia002** [comment]
OK, I have updated the code, PTAL @XiShanYongYe-Chang
